ccvm_simulators/solvers/dl_solver.py

```diff
@@ -444,3 +470,288 @@ def solve(
             solution.evolution_file = evolution_file
 
         return solution
+
+    def __call__(
```

Comment:
I have some concerns about how we are using `__call__()` here.
* Code duplication: A lot of this code is duplicated from the `solve` method, which means there are extra places to update the code if we decide to modify the logic
* User expectation: As a user, I'd expect a the `__call__()` method of a solver object to call the existing `solve()` function, not a variation of it
* Extensibility: If we choose to support more types of algorithms, it will become even more confusing which algorithm is being called when you call the solver instance directly

These concerns could be addressed a couple different ways:
- add an "algorithm" enum to `solve()`, change that logic to support the ADAM algorithm, and make `__call__()` call `solve()` under the hood
- Subclass each solver with an ADAM type of that solver

The second approach would probably be faster to implement, since you just need to override two functions (solve and _calculate_gradients_boxqp), but if conceptually they are the same kind of solver then the first approach would make more sense.

Reply:
Regarding your concern about `__call__()`, my propol as follows:
- Rename `solve()` as `solve_original()` or something more appropriate name
- Rename `__call__()` as `solve_adam()` 
- We have an upcoming method called `solve_sde(...)` using `PyTorch`
- Define a new `__call__()` something like 
```
def __call__(self, solver_type, ...):
    if solver_type=='adam':
        return self.solve_adam(...)
   elif solver_type=='sde':
        return self.solve_sde(...)
   else:
        return self.solve_original(...)
```
We can discuss a bit tomorrow.

Reply:
Sounds great! I like this idea.

Reply:
As suggested now the `__call__()` method for each solver can capable of running desired solver methods. Currently we have solve method for "original" and "adam". Here is the summary of changes:

1. The original method `solve()` is replaced with protected form i.e. `_solve()`. In that case it can only be executed in the `__call__()` method. Docstring is updated. 
2. The adam method is called `_solve_adam()`. Similar to original method, it can be called inside `__call__()`. Doctstring is updated.
3. To address 1 and 2 `CCVMSolver` is updated accordingly
4. The demo scripts in `examples/` were updated and redundant scripts were removed.
5. The readme file in `examples/` is updated correspondingly

Reply:
Beautiful 👌